### PR TITLE
Add tooltip to show levels required for next combat level

### LINKED ISSUE #448: Add levels required for next combat level
Put a tooltip on the "Combat level: XX.X" text in the combat tab that displays how many levels in each combat skill it would take to go up by one combat level.

Related to this, the same functionality could be added to the hiscores panel when you click combat level.

### PR DESCRIPTION
**Pull request by bmid**
# Changes and new additions
* Add new final int under `Experience.java` for maximum combat level
* Add overlay that will calculate which skills and how many levels are necessary to increase combat
* Add new configuration page for the combat level plugin
* Create tests to ensure calculations are accurate

# Images
image

image
The tooltip will leave out anything you are unable to gain a combat level through training.

# Notes
* Previously was under PR #3744, but issues got in the way as RL updated and couldn't get solved.
* Will close #448

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+		return (int) Math.ceil((end - dhp - start) / 0.4875);
+	}
+
+	/***
+	 * Corrects how many levels you need to level up combat through prayer.
+	 * @param level	current prayer level
+	 * @param need	needed prayer level calculated by calcLevels(...)
+	 * @return	a corrected number to increase combat through prayer only
+	 */
+	static int correctPrayer(int level, int need)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: Mark this with `@VisibleForTesting`

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+	/***
+	 * Calculate skill levels required for increasing combat level, meant
+	 * ONLY for Ranged and Magic skills.
+	 * @param start	either the current ranged or magic level
+	 * @param end	ending value (combat level + 1)
+	 * @param dhp	defence, hitpoints, and prayer; this is the initial calculated "base" value
+	 * @return	levels required for a specific skill to level up combat
+	 */
+	static int calcLevelsRM(double start, int end, double dhp)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: Mark this with `@VisibleForTesting`

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+	/***
+	 * Calculate skill levels required for increasing combat level, meant
+	 * for all combat skills besides ranged and magic.
+	 * @param start	initial value
+	 * @param end	ending value (combat level + 1)
+	 * @param multiple	how much adding one skill level will change combat
+	 * @return	levels required for a specific skill to level up combat
+	 */
+	static int calcLevels(double start, int end, double multiple)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: Mark this with `@VisibleForTesting`

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+	@Inject
+	private TooltipManager tooltipManager;
+
+	@Override
+	public Dimension render(Graphics2D graphics)
+	{
+		if (!config.showLevelsUntil()
+			|| client.getLocalPlayer().getCombatLevel() == Experience.MAX_COMBAT_LEVEL
+			|| client.getVar(VarClientInt.CURRENT_TAB) != TabIndex.ATTACK_STYLES.getValue())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: Why is this necessary? just check for existence and visibility of the combat widget

File: runelite-api/src/main/java/net/runelite/api/Experience.java [hunk 5]
```diff
@@ -42,6 +42,11 @@
 	public static final int MAX_VIRT_LEVEL = 126;
 	public static final int MAX_SKILL_XP = 200_000_000;
 
+	/**
+	 * The maximum possible combat level.
+	 */
+	public static final int MAX_COMBAT_LEVEL = 126;
```
Comment (deathbeam) [inline runelite-api/src/main/java/net/runelite/api/Experience.java]: There is already const named MAX_VIRT_LEVEL in this file

Comment (bmiddle) [reply runelite-api/src/main/java/net/runelite/api/Experience.java]: MAX_VIRT_LEVEL is for every trainable skill. If a new combat skill was released, MAX_COMBAT_LEVEL would go probably go up. Should I still remove it?

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java (hunk 4 above)
Comment (bmiddle) [reply runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: If I replace this conditional with that check, do you want me to delete the enum file `TabIndex.java`?

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: yeah probably as it would not be necessary (also the varbit addition)

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+import net.runelite.client.ui.overlay.tooltip.Tooltip;
+import net.runelite.client.ui.overlay.tooltip.TooltipManager;
+import net.runelite.client.util.ColorUtil;
+import javax.inject.Inject;
+import java.awt.Color;
+import java.awt.Dimension;
+import java.awt.Graphics2D;
+import java.awt.Rectangle;
+
+public class CombatLevelOverlay extends Overlay
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: remove public modifier

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+{
+	public static final Color COMBAT_LEVEL_COLOUR = new Color(0xff981f);
+
+	@Inject
+	private Client client;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: Inject stuff through constructor (`@Inject private CombatLevelOverlay(...)`).

Comment (bmiddle) [reply runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: I updated it, I believe I did it correctly now...

File: runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java [hunk 8]
```diff
@@ -40,18 +40,21 @@
...
 import java.awt.Rectangle;
 
-public class CombatLevelOverlay extends Overlay
+class CombatLevelOverlay extends Overlay
 {
-	public static final Color COMBAT_LEVEL_COLOUR = new Color(0xff981f);
+	private static final Color COMBAT_LEVEL_COLOUR = new Color(0xff981f);
 
-	@Inject
 	private Client client;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/combatlevel/CombatLevelOverlay.java]: also mark the fields final

**Comment by bmiddle** [comment]
Is there anything else I should add or fix up in this PR?
